src/Expander.hs

```diff
@@ -551,52 +563,56 @@ initializeKernel = do
+            linkDeclValidity vp (ScopeSet.singleScopeAtPhase sc p)
         )
       , ( "example"
-        , \ sc dest pdest stx -> do
-            p <- currentPhase
+        , \ dest vp stx -> do
             Stx _ _ (_ :: Syntax, expr) <- mustHaveEntries stx
             exprDest <- liftIO $ newSplitCorePtr
             sch <- liftIO newSchemePtr
-            linkDecl dest (Example (view (unSyntax . stxSrcLoc) stx) sch exprDest)
+            linkOneDecl dest (Example (view (unSyntax . stxSrcLoc) stx) sch exprDest)
             t <- inTypeBinder do
               t <- Ty . TMetaVar <$> freshMeta
-              forkExpandSyntax (ExprDest t exprDest) (addScope p expr sc)
+              forkExpandSyntax (ExprDest t exprDest) expr
               return t
             forkGeneralizeType exprDest t sch
-            nowValidAt pdest (SpecificPhase p)
+            linkDeclValidity vp mempty
         )
       , ( "import"
          -- TODO Make import spec language extensible and use bindings rather than literals
-        , \sc dest pdest stx -> do
+        , \dest vp stx -> do
             Stx scs loc (_ :: Syntax, toImport) <- mustHaveEntries stx
             spec <- importSpec toImport
             modExports <- getImports spec
+            sc <- freshScope $ T.pack $ "For import at " ++ shortShow (stxLoc stx)
             flip forExports_ modExports $ \p x b -> inPhase p do
               imported <- addRootScope' $ addScope' (Stx scs loc x) sc
               addImportBinding imported b
-            linkDecl dest (Import spec)
-            nowValidAt pdest AllPhases
+            linkOneDecl dest (Import spec)
+            linkDeclValidity vp (ScopeSet.singleUniversalScope sc)
         )
       , ( "export"
-        , \_sc dest pdest stx -> do
+        , \dest vp stx -> do
             Stx _ _ (_, protoSpec) <- mustBeCons stx
             exported <- exportSpec stx protoSpec
-            p <- currentPhase
             es <- getExports exported
             modifyState $ over expanderModuleExports $ (<> es)
-            linkDecl dest (Export exported)
-            nowValidAt pdest (SpecificPhase p)
+            linkOneDecl dest (Export exported)
+            linkDeclValidity vp mempty
         )
       , ( "meta"
-        , \sc dest pdest stx -> do
-            Stx _ _ (_ :: Syntax, subDecl) <- mustHaveEntries stx
-            subDest <- liftIO newDeclPtr
-            linkDecl dest (Meta subDest)
+        , \dest vp stx -> do
+            (_ :: Syntax, subDecls) <- mustHaveShape stx
+            subDest <- newDeclTreePtr
+            linkOneDecl dest (Meta subDest)
             inEarlierPhase $
-              forkExpandOneDecl subDest sc pdest =<< addRootScope subDecl
+              expandDeclForms subDest mempty vp =<< addRootScope subDecls
+        )
+      , ( "group"
+        , \dest vp stx -> do
+            (_ :: Syntax, decls) <- mustHaveShape stx
+            expandDeclForms dest mempty vp decls
```

Comment:
Turns out it was easier to implement `group` directly than to implement it in terms of `no-decls` and `decls`.

Reply:
Nice!